### PR TITLE
[WIP] Check yield signal every 5s in generated project method

### PR DESCRIPTION
**Pull request by nezihyigitbasi**
As reported in the slack channel checking the yield signal at every row hits the performance. I experimented with the simple idea of checking the yield signal every `5s` (also tried `1s`, but it didn't perform as well). I didn't explore the more complex idea of profiling the project method first and then doing tricks after that, which would make the code generation logic pretty complex (maybe we can get better performance than this simple change though, we don't know).

I ran tpch q1, q10, and q15, which I arbitrarily picked, and I used a scale factor 100. Below are the improvements/savings (*none* of the tests resulted in worse performance, which is nice) in aggregate CPU time spent in the `ScanFilterAndProject` operators in these queries. When there are multiple numbers for a query it means that there are multiple stages with `ScanFilterAndProject` operators (these numbers are aggregated stage-wide). Also, here's the generated project method with this change: https://gist.github.com/nezihyigitbasi/b015a718e49280ed2079e14b2946e1a1

- **q1:** 4%
- **q11:** 20%, 17%, 31%, 19%
- **q15:** 19%, 1%, 1%

Overall, the results seem promising, and it may be worth applying this change. Curious to hear other thoughts.

[Update: tpch q11 is used, not q10]

### CODE REVIEW

**Comment by sopel39** [comment]
Nice investigation! Do you know how those improvements relate to overall query performance?
I wonder what we could do to capture such regressions in the future? Maybe collect basic operator CPU metrics for predefined queries and visualize trendlines?

**Comment by martint** [comment]
This is still calling System.nanoTime for each position+column processed, no? It's worth noting that each call to that function takes around 35ns (based on measurements I've done in the past).

File: presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java [hunk 1]
```diff
@@ -319,18 +338,22 @@ private static MethodDefinition generateProcessMethod(
...
 
-    private static BytecodeBlock generateCheckYieldBlock(Variable thisVariable, Variable index, FieldDefinition yieldSignal, FieldDefinition nextIndexOrPosition)
+    private BytecodeBlock generateCheckYieldBlock(Variable thisVariable, Variable start, Variable index, FieldDefinition yieldSignal, FieldDefinition nextIndexOrPosition)
     {
         return new BytecodeBlock()
-                .comment("if (yieldSignal.isSet())")
+                .comment("if (System.nanoTime() - start > yieldCheckPeriodNanos)")
                 .append(new IfStatement()
-                        .condition(thisVariable.getField(yieldSignal).invoke("isSet", boolean.class))
+                        .condition(greaterThan(subtract(invokeStatic(System.class, "nanoTime", long.class), start), constantLong(yieldCheckPeriodNanos)))
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java]: you can just and these together with `io.airlift.bytecode.expression.BytecodeExpressions#and`

File: presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java [hunk 2]
```diff
@@ -319,18 +338,22 @@ private static MethodDefinition generateProcessMethod(
         return method;
     }
 
-    private static BytecodeBlock generateCheckYieldBlock(Variable thisVariable, Variable index, FieldDefinition yieldSignal, FieldDefinition nextIndexOrPosition)
+    private BytecodeBlock generateCheckYieldBlock(Variable thisVariable, Variable start, Variable index, FieldDefinition yieldSignal, FieldDefinition nextIndexOrPosition)
     {
         return new BytecodeBlock()
-                .comment("if (yieldSignal.isSet())")
+                .comment("if (System.nanoTime() - start > yieldCheckPeriodNanos)")
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java]: Maybe I'm missing it, but where is start updated?

**Comment by nezihyigitbasi** [comment]
> This is still calling System.nanoTime for each position+column processed, no? It's worth noting that each call to that function takes around 35ns (based on measurements I've done in the past).

@martint you are right.

File: presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java (hunk 2 above)
Comment (nezihyigitbasi) [reply presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java]: `start` is initialized on L309.

**Comment by nezihyigitbasi** [comment]
> Nice investigation! Do you know how those improvements relate to overall query performance?
I wonder what we could do to capture such regressions in the future? Maybe collect basic operator CPU metrics for predefined queries and visualize trendlines?

@sopel39 Looking at the overall query performance (elapsed time & total cpu time) I still observe some improvement for all queries, please see the numbers below (with a slightly worse elapsed time for q15. Note that this analysis is far from being scientific and run on my laptop).

Capturing such performance regressions will require looking at performance profiles across changes. One idea is, for each PR, we can have a step that packages Presto master vs. current PR and run tpch (or subset) and get a diff. This may be costly if we do it for each PR though in terms of build times, etc.

Or we can store/keep performance profiles ("baseline") somewhere for tpch queries. We can keep operator level stats as you mention (cpu and memory-related stats) and again we can diff for current PR vs. baseline. Overall, automating the process of comparing the performance of a PR against some baseline requires some work, but I think it will be valuable in the long run. 

In the short term, for changes to the critical part of the engine we can require proper benchmarks & results before merging them -- we already do this for the most part, but since we still see changes like this one getting in, there is probably some room for improvement.

```
q11: 
    - elapsed time before: 2.87m after: 1.69m
    - total cpu time: 8.19m after: 6.18m
q1:
    - elapsed time before: 4.56m after: 4.00m
    - total cpu time: 23.83m after: 23.01m

q15:
    - elapsed time before: 3.51m after: 3.85m
    - total cpu time: 19.05m after: 18.99m
```

**Comment by findepi** [comment]
> This is still calling System.nanoTime for each position+column processed

> I didn't explore the more complex idea of profiling the project method first and then doing tricks after that, which would make the code generation logic pretty complex

What about having this profiling in non-generated class?
In the generated code you would just call some API. I.e. only equivalent to the following would need to be generated:

```
if (expressionProfile.shouldCheckYield(currentPosition, currentColumn) && this.yieldSignal.isSet()) {
   ....
}
```
`expressionProfile` could keep internally all the timing state and decide when to consult the clock

**Comment by sopel39** [comment]
> Or we can store/keep performance profiles ("baseline") somewhere for tpch queries. We can keep operator level stats as you mention (cpu and memory-related stats) and again we can diff for current PR vs. baseline.

Keep in mind that individual benchmarks have a high variability, so pair comparisons might show false negatives. It is more useful to compare performance trends and create alerts on a longer timeline.

**Comment by sopel39** [comment]
I wonder if a generic logic shouldn't be part of DriverYielsSignal class so that it is transparent for operators and can be reused seamlessly. There are other places which could take benefit of less aggressive yield check.

**Comment by dain** [comment]
@nezihyigitbasi, I'd like to understand why this is faster.  The current yield signal is an AtomicBoolean, so the cost should be similar to a volatile read which is ~1ns according to this benchmark from Shipilev (https://shipilev.net/blog/2014/all-accesses-are-atomic/).  He also has a blog on some performance problems you can get in code like DriverYieldSignal and how to work around them (https://shipilev.net/blog/2015/faster-atomic-fu/).

**Comment by nezihyigitbasi** [comment]
If I speculate for a second, I would guess that atomic boolean reads may be negatively impacting the cache due to invalidations, to understand the root cause more investigation is needed.

I will also experiment with the profiling approach and see how that performs.

**Comment by nezihyigitbasi** [comment]
I will open another PR later (exploring the profiler approach) as there are problems with this benchmark. More investigation is needed.

**Pull request closed without merging**